src/components/_global/BalForm/useForm.ts

```diff
@@ -0,0 +1,158 @@
+import { RuleFunction, Rules } from '@/types';
+import { get, set } from 'lodash';
+import { onBeforeMount, reactive, toRefs } from 'vue';
+
+/**
+ * TYPES
+ */
+type UseFormRequest<T> = {
+  name: string;
+  defaultValues?: Partial<T>;
+};
+
+export type FormState<T> = {
+  name: string;
+  values: T;
+  errors: { [key in keyof T]: string[] };
+  rules: { [key in keyof T]: RuleFunction[] | undefined };
+  defaultValues: T;
+};
+
+export type FormInstance<T> = FormState<T> & {
+  register: (name: string, rules?: Rules) => any;
+  onChange: (name: string, value: any) => void;
+  handleSubmit: (submit: () => void) => void;
+};
+
+function validateAgainst(value: any, rules?: RuleFunction[]) {
+  if (rules?.length) {
+    const failedRules = rules
+      .map(rule => {
+        const pass = rule(value);
+        if (typeof pass === 'string') {
+          return pass;
+        }
+      })
+      .filter(result => result !== undefined);
+    return failedRules;
+  }
+  return [];
+}
+
+/**
+ * GLOBAL
+ */
+const formCache = reactive<Record<string, FormInstance<any>>>({});
+export function useFormCache(name: string) {
+  return formCache[name];
+}
+
+export const FormContextSymbol = Symbol('FORM_CONTEXT');
```

Comment:
Is there a chance of clashing here if there are multiple forms on the same page? It's not clear to me what this symbol is for in useFormContext.